### DISCUSSION TITLE
Creating Type-safe ClsService using custom Module

### DISCUSSION THREAD

**Discussion by MaheeGamage** [Papooch/nestjs-cls#70] [Q&A] [answered]
Can someone explain to me how to use this module that explained in README.md file
https://github.com/Papooch/nestjs-cls#type-safety-and-type-inference

```ts
export class MyClsService extends ClsService<MyClsStore> {}

@Module({
    imports: [ClsModule.forFeature()],
    providers: [{
        provide: MyClsService,
        useExisting: ClsService
    }],
    exports: [MyClsService]
})
class MyClsModule {}
```

I need to know after creating a custom `MyClsModule` how to add this module to the application.

Currently I use `nestjs-cls` in following manner
```ts
@Module({
  imports: [
    // Register the ClsModule,
    ClsModule.forRoot({
      middleware: {
        // automatically mount the
        // ClsMiddleware for all routes
        mount: true,
        // and use the setup method to
        // provide default store values.
        setup: (cls, req) => {
          cls.set('userId', req.headers['x-user-id']);
        },
      },
    }),
  ],
  providers: [CatService],
  controllers: [CatController],
})
export class AppModule {}
```

**Comment by Papooch** [comment, marked as answer]
You just inject `MyClsService` the same way you would inject `ClsService`. It will be typed with your `MyClsStore`. Of course, `MyClsModule` must be imported in the module where `MyClsService` is needed.

You do that _in addition_ to registering it in for-root. With this method, you're basically just aliasing `ClsService` to `MyClsService` to get type inference, but the same `ClsService` is still used underneath.

If you want type safety even in the `setup` method, you can type it like so:
```ts
setup: (cls: MyClsService, req) => {
  cls.set('userId', req.headers['x-user-id']);
},
```

**Reply by Papooch** [reply]
An alternative to using the wrapper module is to type your `ClsService` injects as `ClsService<MyClsStore>`, and therefore:
```ts
setup: (cls: ClsService<MyClsStore>, req) => {
  cls.set('userId', req.headers['x-user-id']);
},
```
This way you can avoid having to create the alias class, but have to use the generic type everywhere.
